[Draft] Improved tree shaking code

EloB opened the pull request:
# Draft

This is a draft to improve tree shaking and dead code elimination when bundling `typia`.

I created this PR to get early feedback.

When I have more time I will provide an example how I use this combined with a esbuild plugin to bundle typia automatically and that it tree shakes everything that it doesn't need.

## Code changes:
- Replaced `namespace name {}` with `export * as namespaceName from "./name"`. Improves dead code elimination.
- Added `/** @__PURE__ */` to some exports that has `Object.assign`.


## Example
I bundled the code below with esbuild and I was able to compress the bundle from 69827 characters down to 5804 (**-92%**). We also need to add `"sideEffects": false` to `ret` and `randexp` package.json files. I've made a PR to `randexp` but have to make another for `ret`.
```ts
import { IValidation, tags, validate } from "typia";

const res: IValidation<IMember> = validate<IMember>({
  id: 5, // wrong, must be string (uuid)
  age: 20.75, // wrong, not integer
  email: "samchon.github@gmail.com",
});

if (!res.success) console.log(res.errors);

interface IMember {
  id: string & tags.Format<"uuid">;
  email: string & tags.Format<"email">;
  age: number
  & tags.Type<"uint32">
  & tags.ExclusiveMinimum<19>
  & tags.Maximum<100>;
}
```

EloB commented on package.json:
```diff
@@ -15,7 +15,7 @@
     "test": "npm run package:tgz",
     "-------------------------------------------------": "",
-    "build": "rimraf lib && tsc --removeComments --declaration false && tsc --emitDeclarationOnly",
+    "build": "rimraf lib && tsc --declaration false && tsc --emitDeclarationOnly",
```
I include the comments so it doesn't strip `/** @__PURE__ */`.

EloB committed:
feature: improved tree shaking code

samchon reviewed (commented):
Thanks for contribution, but bundled size is increased from 11.2kb to 11.5kb.

  - https://bundlephobia.com/package/typia@5.3.11
  - https://bundlephobia.com/package/typia@5.3.12-dev.20240118

However, it seems like you're expert at the package bundling. Looking your PR, I got another idea of bundling diet way through the `require()` function. What about changing `typia` to generate `require()` function instead of using the namespace like below? 

If this way works, it may possible to diet near to the 0 byte, so that don't need to consider tree shaking problem.

```typescript
// CURRENT
const $guard = typia_1.default.json.createAssertStringify.guard;

// HOW ABOUT THIS ONE?
const $guard = require("typia/lib/functional/$guard").$guard;
```

EloB commented on src/functional/$ProtobufReader.ts:
```diff
@@ -189,7 +189,7 @@ export class $ProtobufReader {
-const N7F = BigInt(0x7f);
-const N80 = BigInt(0x80);
+const utf8 = /** @__PURE__ */ new TextDecoder();
+const N01 = /** @__PURE__ */ BigInt(0x01);
```
I think these could be change to not execute functions in the root scope.

```js
const N01 = 0x01n;
const N7F = 0x7fn;
const N80 = 0x80n;
```

EloB replied:
The syntax for `BigInt` without using `BigInt(value)` is to add an "n" (`1n === BigInt(1)`) in the end of a integer.

EloB commented on src/transformers/CallExpressionTransformer.ts:
```diff
@@ -92,38 +92,38 @@ import { ReflectMetadataTransformer } from "./features/reflect/ReflectMetadataTr
-      const declaration: ts.Declaration | undefined =
-        project.checker.getResolvedSignature(expression)?.declaration;
-      if (!declaration) return expression;
+      (expression: ts.CallExpression): ts.Expression | null => {
```
I will revert these. My editor must had formatted them.

samchon committed:
Refactor exported functions.

Changed exported function not to show the namespace nested functions' definitions.

EloB committed:
fix: correct pure reference for Object.assign

samchon reviewed (approved):
Good, I'd missed somethinig.

Now, may I do publish?

samchon reviewed (approved):
Everything working fine. Thanks for contribution and waiting for your esbuild plugin.
